src/core/TESTING.md

```diff
@@ -62,8 +62,257 @@ test('my test', async () => {
 ### Strategies for specific Core APIs
 
 #### HTTP Routes
+The HTTP API interface is another public contract of Kibana. Although not every Kibana endpoint is for external use. Evaluating the required level of test coverage for an HTTP resource, make your judgment based on the fact whether an endpoint is considered to be public or private. Public API is expected to have a higher level of test coverage.
+Public API tests should cover **observable behavior** of the system. Therefore they should be close to the real user interactions as much as possible. Ideally, to use HTTP requests to communicate with the Kibana server.
+
+##### Integration tests
+Depending on the number of external dependencies, you can consider implementing several high-level integration tests. They would work as a set of [smoke tests](https://en.wikipedia.org/wiki/Smoke_testing_(software)) for the most important functionality.
+Subjects for tests:
+- authenticated / unauthenticated access to an endpoint.
+- endpoint validation (params, query, body).
+- main business logic.
+
+###### Functional Test Runner
+If your plugin leverages elasticsearch server to store data and supports additional configuration, you can utilize Functional Test Runner(FTR) to add integration tests. 
+FTR bootstraps the real elasticsearch instance and runs Kibana against it.
+Pros:
+- runs the whole Elastic stack
+- tests cross-plugin integration
+- emulates a real user interaction with the stack
+- allows adjusting config values
+Cons:
+- slow start
+- hard to debug
+- brittle tests
+
+Imagine that `myPlugin` provides a set of endpoints to store & retrieve user data:
+```ts
+router.post(
+  {
+    path: '/myPlugin/something',
+    validate: {
+      body: schema.object({
+        text: schema.string({ maxLength: 100 }),
+      }),
+    },
+  },
+  async (context, request, response) => {
+    const savedObjectsClient = context.core.savedObjects.client;
+    const { id } = await savedObjectsClient.update('myPlugin-type', 'myPlugin', {
+      something: request.body.text,
+    });
+    return response.ok({ body: { id } });
+  }
+);
+
+router.get(
+  {
+    path: '/myPlugin/something/{id}',
+    validate: {
+      params: schema.object({
+        id: schema.string(),
+      }),
+    },
+  },
+  async (context, request, response) => {
+    const savedObjectsClient = context.core.savedObjects.client;
+    const { attributes } = await savedObjectsClient.get('myPlugin-type', request.params.id);
```

Comment:
NIT: should we add try/catch around the SO call to returns a 'proper' error from the handler in the exemple?